crates/red_knot_python_semantic/src/types/context.rs

```diff
@@ -221,3 +201,268 @@ pub(crate) enum InNoTypeCheck {
+/// than `DiagnosticId::Lint`, then they should use the more general
+/// `InferContext::report` API. But note that this API will not take rule
+/// selection or suppressions into account.
+pub(super) struct LintReporter<'db, 'ctx> {
+    /// The typing context.
+    ctx: &'ctx InferContext<'db>,
+    /// The diagnostic that we want to report.
+    ///
+    /// This is always `Some` until the `Drop` impl.
+    diag: Option<Diagnostic>,
+    /// The lint ID. Stored here because it doesn't
+    /// seem easily derivable from `DiagnosticId` and
+    /// because we need it for looking up suppressions.
+    lint_id: LintId,
+}
+
+impl LintReporter<'_, '_> {
+    /// Return a mutable borrow of the diagnostic on this reporter.
+    ///
+    /// Callers may mutate the diagnostic to add new sub-diagnostics
+    /// or annotations.
+    ///
+    /// The diagnostic is added to the typing context, if appropriate,
+    /// when this reporter is dropped.
+    pub(super) fn diagnostic(&mut self) -> &mut Diagnostic {
+        // OK because `self.diag` is only `None` within `Drop`.
+        self.diag.as_mut().unwrap()
+    }
+}
+
+/// Finishes use of this reporter.
+///
+/// This will add the lint as a diagnostic to the typing context if
+/// appropriate. The diagnostic may be skipped, for example, if there is a
+/// relevant suppression.
+impl Drop for LintReporter<'_, '_> {
+    fn drop(&mut self) {
+        // OK because the only way `self.diag` is `None`
+        // is via this impl, which can only run at most
+        // once.
+        let diag = self.diag.take().unwrap();
+        let Some(span) = diag.primary_span() else {
+            self.ctx.diagnostics.borrow_mut().push(diag);
+            return;
+        };
+        let Some(range) = span.range() else {
+            self.ctx.diagnostics.borrow_mut().push(diag);
+            return;
+        };
+        // Two things to note here.
+        //
+        // First is that we use `span.file()` here to find suppressions
+        // and *not* self.reporter.ctx.file` (as this code used to do).
+        // The reasoning for this is that the suppression ought to be
+        // based on the primary span of the diagnostic itself, and not
+        // the file that happens to be checked currently. It seems likely
+        // that in most (all?) cases, these will be the same. But in a
+        // hypothetical case where it isn't, this seems like the more
+        // sensible option.
+        //
+        // Second is that we are only checking suppressions here when
+        // a range is present. But it seems like we could check
+        // suppressions even when a range isn't present, since they
+        // could be file-level suppressions. However, it's not clear,
+        // in practice, when this matters. In particular, it is generally
+        // expected that most (all?) lint diagnostics will come with at
+        // least one primary span.
```

Comment:
Yeah, I think we should just make this mandatory. A missing range means that the diagnostic isn't actionable. You don't even know from which file it comes from.